recipes/m4/all/conanfile.py

```diff
@@ -0,0 +1,99 @@
+from conans import ConanFile, tools, AutoToolsBuildEnvironment
+from contextlib import contextmanager
+import os
+
+
+class M4Conan(ConanFile):
+    name = "m4"
+    description = "GNU M4 is an implementation of the traditional Unix macro processor"
+    topics = ("conan", "m4", "macro", "macro processor")
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://www.gnu.org/software/m4/"
+    license = "GPL-3.0-only"
+    exports_sources = ["patches/*.patch"]
+    settings = "os_build", "arch_build", "compiler"
+
+    _autotools = None
+    _source_subfolder = "source_subfolder"
+    _build_subfolder = "build_subfolder"
+
+    @property
+    def _is_msvc(self):
+        return self.settings.compiler == "Visual Studio"
+
+    @property
+    def _is_clang(self):
+        return str(self.settings.compiler).endswith("clang")
+
+    def build_requirements(self):
+        if tools.os_info.is_windows and "CONAN_BASH_PATH" not in os.environ and \
+                tools.os_info.detect_windows_subsystem() != "msys2":
+            self.build_requires("msys2/20190524")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename("m4-" + self.version, self._source_subfolder)
+
+    def _configure_autotools(self):
+        if self._autotools:
+            return self._autotools
+        conf_args = []
+        self._autotools = AutoToolsBuildEnvironment(self, win_bash=tools.os_info.is_windows)
+        self._autotools.configure(args=conf_args, configure_dir=self._source_subfolder)
+        return self._autotools
+
+    @contextmanager
+    def _build_context(self):
+        env = {}
+        if self.settings.compiler == "Visual Studio":
+            with tools.vcvars(self.settings):
+                env.update({
+                    "AR": "{}/build-aux/ar-lib lib".format(tools.unix_path(self._source_subfolder)),
+                    "CC": "cl -nologo",
+                    "CXX": "cl -nologo",
+                    "LD": "link",
+                    "NM": "dumpbin -symbols",
+                    "OBJDUMP": ":",
+                    "RANLIB": ":",
+                    "STRIP": ":",
+                })
+                with tools.environment_append(env):
+                    yield
+        else:
+            if "clang" in str(self.settings.compiler):
+                env["CFLAGS"] = "-rtlib=compiler-rt"
+            with tools.environment_append(env):
+                yield
+
+    def _patch_sources(self):
+        for patch in self.conan_data["patches"][self.version]:
+            self.output.info("applying patch \"{}\"".format(patch["patch_file"]))
```

Comment:
not sure if this info is required. The post_export hook shows all patches which will be applied.